core/trace/src/main/scala/org/typelevel/otel4s/trace/SpanContext.scala

```diff
@@ -104,14 +122,147 @@ object SpanContext {
   }
 
   val invalid: SpanContext =
-    new SpanContext {
-      val traceIdHex: String = TraceId.InvalidHex
-      val traceId: ByteVector = ByteVector.fromValidHex(traceIdHex)
-      val spanIdHex: String = SpanId.InvalidHex
-      val spanId: ByteVector = ByteVector.fromValidHex(spanIdHex)
-      val traceFlags: TraceFlags = TraceFlags.Default
-      val traceState: TraceState = TraceState.empty
-      val isValid: Boolean = false
-      val isRemote: Boolean = false
+    SpanContextImpl(
+      traceId = TraceId.Invalid,
+      traceIdHex = TraceId.Invalid.toHex,
+      spanId = SpanId.Invalid,
+      spanIdHex = SpanId.Invalid.toHex,
+      traceFlags = TraceFlags.Default,
+      traceState = TraceState.empty,
+      isRemote = false,
+      isValid = false
+    )
+
+  /** Creates a new [[SpanContext]] with the given identifiers and options.
+    *
+    * If the `traceId` or the `spanId` are invalid (ie. do not conform to the
+    * requirements for hexadecimal ids of the appropriate lengths), both will be
+    * replaced with the standard "invalid" versions (i.e. all '0's).
+    *
+    * @param traceId
+    *   the trace identifier of the span context
+    *
+    * @param spanId
+    *   the span identifier of the span context
+    *
+    * @param traceFlags
+    *   the trace flags of the span context
+    *
+    * @param traceState
+    *   the trace state of the span context
+    *
+    * @param remote
+    *   whether the span is propagated from the remote parent or not
+    */
+  def apply(
+      traceId: ByteVector,
+      spanId: ByteVector,
+      traceFlags: TraceFlags,
+      traceState: TraceState,
+      remote: Boolean
+  ): SpanContext =
+    if (TraceId.isValid(traceId) && SpanId.isValid(spanId)) {
+      createInternal(
+        traceId = traceId,
+        spanId = spanId,
+        traceFlags = traceFlags,
+        traceState = traceState,
+        remote = remote,
+        isValid = true
+      )
+    } else {
+      createInternal(
+        traceId = TraceId.Invalid,
+        spanId = SpanId.Invalid,
+        traceFlags = traceFlags,
+        traceState = traceState,
+        remote = remote,
+        isValid = false
+      )
     }
+
+  implicit val spanContextHash: Hash[SpanContext] =
+    Hash.by { ctx =>
+      (
+        ctx.traceIdHex,
+        ctx.spanIdHex,
```

Comment:
any particular reason to use the hex `String`s for the `Hash` rather than the `ByteVector`s?

Reply:
Not really. We can use `ctx.traceId` and `ctx.spanId` instead.

Reply:
There is no `Hash[ByteVector]` instance, unfortunately.

Reply:
Oh, we can inline a quickie one with `Hash.fromUniversalHashCode()`. I think instances might live in `scodec-cats` but not worth dragging that in ...

Reply:
Is there any benefit of using `traceIdHex` vs `traceId`? Since we treat `traceId` as a 'primary' property, we should use it in hashing too, right?

Reply:
> Since we treat `traceId` as a 'primary' property, we should use it in hashing too, right?

That's my thinking as well.